File: classify/th/cnn-sentence.lua
--[[

Efficient Implementation of Static CMOT.  This model doesnt use any
LookupTable, doesnt fine-tune embeddings, they are loaded as feature vectors.
This model is therefore fairly simple to implement without a deep learning 
frameworks, assuming a good SGD implementation.

]]--

require 'nn'
require 'xlua'
require 'optim'
require 'classutils'
require 'data'
require 'train'
require 'emb'
torch.setdefaulttensortype('torch.FloatTensor')

-----------------------------------------------------
-- Defaults if you dont specify args
-- Note that some arguments depend on which optim
-- is selected, and may be unused for others.  Try to
-- provide reasonable args for any algorithm a user selects
-----------------------------------------------------
DEF_TSF = './data/TREC.train.all'
DEF_ESF = './data/TREC.test.all'
DEF_BATCHSZ = 10
DEF_OPTIM = 'sgd'
DEF_ETA = 0.01
DEF_MOM = 0.0
DEF_DECAY = 1e-9
DEF_DROP = 0.5
DEF_MXLEN = 100
DEF_ESZ = 300
DEF_CMOTSZ = 200
DEF_HSZ = -1 -- No additional projection layer
DEF_EMBED = './data/GoogleNews-vectors-negative300.bin'
DEF_FILE_OUT = './cnn-sentence.model';
DEF_FSZ = '{5}'
DEF_PATIENCE = 20
DEF_EPOCHS = 200
DEF_PROC = 'gpu'
DEF_CACTIVE = 'relu'
DEF_HACTIVE = 'none'

----------------------------------------------
-- Make a Softmax output CMOT with Dropout
----------------------------------------------
function createModel(dsz, cmotsz, cactive, hsz, hactive, filts, gpu, nc, pdrop)
    local seq = nn.Sequential()

    local concat = nn.Concat(2)
    for i=1,#filts do
       local filtsz = filts[i]
       print('Creating filter of size ' .. filtsz)
       local subseq = nn.Sequential()
       subseq:add(nn.TemporalConvolution(dsz, cmotsz, filtsz))
       subseq:add(activationFor(cactive))
       subseq:add(nn.Max(2))
       subseq:add(nn.Dropout(pdrop))
       concat:add(subseq)
    end
    seq:add(concat)
    if hsz > 0 then
       seq:add(nn.Linear(#filts * cmotsz, hsz))
       seq:add(activationFor(hactive))
       seq:add(nn.Linear(hsz, nc))
    else
       seq:add(nn.Linear(#filts * cmotsz, nc))
    end
    seq:add(nn.LogSoftMax())
    return gpu and seq:cuda() or seq
end

--------------------------
-- Command line handling
--------------------------
local cmd = torch.CmdLine()
cmd:text('Parameters for Static CMOT Network')
cmd:text()
cmd:text('Options:')
cmd:option('-save', DEF_FILE_OUT, 'Save model to')
cmd:option('-embed', DEF_EMBED, 'Word2Vec embeddings')
cmd:option('-cembed', 'none', 'Char embeddings')
cmd:option('-eta', DEF_ETA, 'Initial learning rate')
cmd:option('-optim', DEF_OPTIM, 'Optimization method (sgd|adagrad|adadelta|adam)')
cmd:option('-decay', DEF_DECAY, 'Weight decay')
cmd:option('-dropout', DEF_DROP, 'Dropout prob')
cmd:option('-mom', DEF_MOM, 'Momentum for SGD')
cmd:option('-train', DEF_TSF, 'Training file')
cmd:option('-eval', DEF_ESF, 'Testing file')
cmd:option('-epochs', DEF_EPOCHS, 'Number of epochs')
cmd:option('-proc', DEF_PROC, 'Backend (gpu|cpu)')
cmd:option('-batchsz', DEF_BATCHSZ, 'Batch size')
cmd:option('-mxlen', DEF_MXLEN, 'Max number of tokens to use')
cmd:option('-patience', DEF_PATIENCE, 'How many failures to improve until quitting')
cmd:option('-hsz', DEF_HSZ, 'Depth of additional hidden layer')
cmd:option('-cmotsz', DEF_CMOTSZ, 'Depth of convolutional/max-over-time output') 
cmd:option('-cactive', DEF_CACTIVE, 'Activation function following conv')
cmd:option('-filtsz', DEF_FSZ, 'Convolution filter width')
cmd:option('-lower', false, 'Lower case words')
DEF_CACTIVE = 'relu'
DEF_HACTIVE = 'relu'

local opt = cmd:parse(arg)
opt.filtsz = loadstring("return " .. opt.filtsz)()
----------------------------------------
-- Optimization
----------------------------------------
state, optmeth = optimMethod(opt)

----------------------------------------
-- Processing on GPU or CPU
----------------------------------------
opt.gpu = false
if opt.proc == 'gpu' then
   opt.gpu = true
   require 'cutorch'
   require 'cunn'
else
   opt.proc = 'cpu'
end
print('Processing on ' .. opt.proc)

---------------------------------------
-- Minibatches
---------------------------------------
print('Using batch size ' .. opt.batchsz)

---------------------------------------
-- Load Word2Vec Model(s)
---------------------------------------
local w2v = Word2VecModel(opt.embed)
print('Loaded word embeddings')
local dsz = w2v.dsz

-- If you want to give us character vectors, we will average and concat
if opt.cembed ~= 'none' then
   opt.w2cv = Word2VecModel(opt.cembed)
   dsz = dsz + opt.w2cv.dsz
   print('Loaded char embeddings')
end

---------------------------------------
-- Load Feature Vectors
---------------------------------------
local f2i = {}
ts,f2i = loadTemporalEmb(opt.train, w2v, f2i, opt)
print('Loaded training data')
es,f2i = loadTemporalEmb(opt.eval, w2v, f2i, opt)
print('Loaded test data')
local i2f = revlut(f2i)
print(f2i)
print(#i2f)

---------------------------------------
-- Build model and criterion
---------------------------------------
local crit = createCrit(opt.gpu, #i2f)
local model = createModel(dsz, opt.cmotsz, opt.cactive, opt.hsz, opt.hactive, opt.filtsz, opt.gpu, #i2f, opt.dropout)


local errmin = 1
local lastImproved = 0

for i=1,opt.epochs do
    print('Training epoch ' .. i)
    confusion = optim.ConfusionMatrix(i2f)
    trainEpoch(crit, model, ts, optmeth, confusion, opt)
    confusion = optim.ConfusionMatrix(i2f)
    local erate = test(model, es, confusion, opt)
    if erate < errmin then
       errmin = erate
       lastImproved = i
       print('Lowest error achieved yet -- writing model')
       saveModel(model, opt.save, opt.gpu)
    end
    if (i - lastImproved) > opt.patience then
       print('Stopping due to persistent failures to improve')
       break
    end
end

print('Highest test acc: ' .. (100 * (1. - errmin)))

File: classify/tf/cnn-sentence.py
import tensorflow as tf
import numpy as np
import w2v
from data import loadTemporalEmb
from model import ConvModelStatic
from data import buildVocab
from data import validSplit
from utils import revlut
from train import Trainer
import time
flags = tf.app.flags
FLAGS = flags.FLAGS
flags.DEFINE_float('eta', 0.001, 'Initial learning rate.')
flags.DEFINE_string('embed', None, 'Word2Vec embeddings file')
flags.DEFINE_string('optim', 'adam', 'Optim method')
flags.DEFINE_float('dropout', 0.5, 'Dropout probability')
flags.DEFINE_string('train', None, 'Training file')
flags.DEFINE_string('valid', None, 'Validation file')
flags.DEFINE_string('test', None, 'Test file')
flags.DEFINE_float('unif', 0.25, 'Initializer bounds for embeddings')
flags.DEFINE_integer('epochs', 25, 'Number of epochs')
flags.DEFINE_integer('batchsz', 50, 'Batch size')
flags.DEFINE_integer('mxlen', 100, 'Max length')
flags.DEFINE_integer('patience', 10, 'Patience')
flags.DEFINE_integer('cmotsz', 100, 'Hidden layer size')
flags.DEFINE_integer('hsz', -1, 'Projection layer size')
flags.DEFINE_string('outdir', 'out', 'Directory to put the output')
flags.DEFINE_string('filtsz', '3,4,5', 'Filter sizes')
flags.DEFINE_boolean('clean', True, 'Do cleaning')
flags.DEFINE_float('valsplit', 0.15, 'Validation split if no valid set')

w2vModel = w2v.Word2VecModel(FLAGS.embed)

f2i = {}
opts = { 'batchsz': FLAGS.batchsz,
         'clean': FLAGS.clean,
         'filtsz': [int(filt) for filt in FLAGS.filtsz.split(',')],
         'mxlen': FLAGS.mxlen }
ts, f2i = loadTemporalEmb(FLAGS.train, w2vModel, f2i, opts)
es, f2i = loadTemporalEmb(FLAGS.test, w2vModel, f2i, opts)
print('Loaded  training data')

if FLAGS.valid is not None:
    print('Using provided validation data')
    vs, f2i = loadTemporalEmb(FLAGS.valid, w2vModel, f2i, opts)
else:
    ts, vs = validSplit(ts, FLAGS.valsplit)
    print('Created validation split')

es, f2i = loadTemporalEmb(FLAGS.test, w2vModel, f2i, opts)
print('Loaded test data')



"""
Train convolutional sentence model
"""

model = ConvModelStatic()

with tf.Graph().as_default():
    sess = tf.Session()
    with sess.as_default():
        model.params(f2i, w2vModel, FLAGS.mxlen, FLAGS.filtsz, FLAGS.cmotsz, FLAGS.hsz)
   
        trainer = Trainer(model, FLAGS.optim, FLAGS.eta)
        train_writer = tf.train.SummaryWriter(FLAGS.outdir + "/train", sess.graph)
        init = tf.initialize_all_variables()
        sess.run(init)
        trainer.prepare(tf.train.Saver())

        max_acc = 0
        last_improved = 0

        for i in range(FLAGS.epochs):
            print('Training epoch %d' % (i+1))
            trainer.train(ts, sess, train_writer, FLAGS.dropout)
            this_acc = trainer.test(vs, sess)
            if this_acc > max_acc:
                max_acc = this_acc
                last_improved = i
                trainer.checkpoint(sess, FLAGS.outdir, "cnn-sentence")

            if (i - last_improved) > FLAGS.patience:
                print('Stopping due to persistent failures to improve')
                break


        print("-----------------------------------------------------")
        print('Highest validation acc %.4f' % max_acc)
        print('=====================================================')
        print('Evaluating best model on test data:')
        print('=====================================================')
        trainer.recover_last_checkpoint(sess, FLAGS.outdir)
        trainer.test(es, sess)

        # Write out model, graph and saver for future inference
        model.save(sess, FLAGS.outdir, 'cnn-sentence')
